lib/datastore/data-store-new.js

```diff
@@ -0,0 +1,175 @@
+import { TypedEmitter } from 'tiny-typed-emitter'
+import { encode, decode } from '@mapeo/schema'
+import MultiCoreIndexer from 'multi-core-indexer'
+import Ram from 'random-access-memory'
+import pDefer from 'p-defer'
+
+/**
+ * @typedef {import('multi-core-indexer').IndexEvents} IndexEvents
+ */
+/**
+ * @typedef {import('@mapeo/schema').MapeoDoc} MapeoDoc
+ */
+/**
+ * @typedef {object} DefaultEmitterEvents
+ * @property {(eventName: keyof IndexEvents, listener: (...args: any[]) => any) => void} newListener
+ * @property {(eventName: keyof IndexEvents, listener: (...args: any[]) => any) => void} removeListener
+ */
+/**
+ * @template T
+ * @template {keyof any} K
+ * @typedef {T extends any ? Omit<T, K> : never} OmitUnion
+ */
+
+const NAMESPACE_SCHEMAS = /** @type {const} */ ({
+  data: ['observation'],
+  auth: [],
+})
+
+/**
+ * @typedef {typeof NAMESPACE_SCHEMAS} NamespaceSchemas
+ */
+
+/**
+ * @template {keyof NamespaceSchemas} TNamespace
+ * @template {NamespaceSchemas[TNamespace][number]} TSchemaName
+ * @extends {TypedEmitter<IndexEvents & DefaultEmitterEvents>}
+ */
+export class DataStore extends TypedEmitter {
+  #coreManager
+  #namespace
+  #writerCore
+  #coreIndexer
+  /** @type {Map<string, import('p-defer').DeferredPromise<void>>} */
+  #pending = new Map()
+
+  /**
+   * @param {object} opts
+   * @param {import('../core-manager/index.js').CoreManager} opts.coreManager
+   * @param {TNamespace} opts.namespace
+   * @param {(entries: MultiCoreIndexer.Entry<'binary'>[]) => Promise<void>} opts.indexEntries
+   */
+  constructor({ coreManager, namespace, indexEntries }) {
+    super()
+    this.#coreManager = coreManager
+    this.#namespace = namespace
+    this.#writerCore = coreManager.getWriterCore(namespace).core
+    const cores = coreManager.getCores(namespace).map((cr) => cr.core)
+    this.#coreIndexer = new MultiCoreIndexer(cores, {
+      storage: (name) => new Ram(name),
+      batch: (entries) => this.#batch(entries, indexEntries),
+    })
+
+    // Forward events from coreIndexer
+    this.on('newListener', (eventName, listener) => {
+      if (['newListener', 'removeListener'].includes(eventName)) return
+      this.#coreIndexer.on(eventName, listener)
+    })
+    this.on('removeListener', (eventName, listener) => {
+      if (['newListener', 'removeListener'].includes(eventName)) return
```

Comment:
extract these events to a constant? i.e. `const LISTENER_EVENTS = ['newListener', 'removeListener']` (you can come up with a better variable name 😄 )

Reply:
I'd prefer not to because I'd like this to be explicit in the code, rather than the indirection of a constant - we'd end up naming it "newListenerAndRemoveListenerEventNames". Making is a constant might be a slight performance improvement, but this is not somewhere that is a bottleneck (events are added rarely).